### ISSUE TITLE
Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage) java.lang.reflect.InvocationTargetException: null

### ISSUE THREAD

**Issue by wangfeng-skymind** [SkymindIO/SKIL-CE#47]
**although skil-env.sh had been set the arguments**

MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage
skil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000
skil.hdfs2modelstorage.basedir=/user/hadoop/modelstorage
skil.hdfs2modelstorage.hadoopuser=root

**however, still show below exception;**

ERROR i.s.s.daemon.ProcessLauncherDaemon - Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage)
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.runMain(ProcessLauncherDaemon.java:522)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.main(ProcessLauncherDaemon.java:1394)
Caused by: java.lang.RuntimeException: Cannot initialize Hdfs2ModelStorage: HDFS path (path to HDFS cluster namenode) not set - set via environment or system property with key "skil.hdfs2modelstorage.hdfsuri"
        at io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage.<init>(Hdfs2ModelStorage.java:71)
        ... 6 common frames omitted
09:10:43.897 [Curator-LeaderSelector-0] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error starting process
java.lang.InterruptedException: null
        at java.lang.Object.wait(Native Method)
        at java.lang.Object.wait(Object.java:502)
        at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1342)
        at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:781)
        at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:691)
        at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:675)
        at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:107)
        at org.apache.curator.framework.imps.CreateBuilderImpl.pathInForeground(CreateBuilderImpl.java:672)
        at org.apache.curator.framework.imps.CreateBuilderImpl.protectedPathInForeground(CreateBuilderImpl.java:453)
        at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:443)
        at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:423)
        at org.apache.curator.framework.imps.CreateBuilderImpl$4.forPath(CreateBuilderImpl.java:331)
        at org.apache.curator.framework.imps.CreateBuilderImpl$4.forPath(CreateBuilderImpl.java:267)
        at io.skymind.skil.daemon.ServiceRegister.allAgents(ServiceRegister.java:273)
        at io.skymind.skil.daemon.ServiceRegister.getAgent(ServiceRegister.java:337)
        at io.skymind.skil.daemon.ServiceRegister.register(ServiceRegister.java:429)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.registerService(ProcessLauncherDaemon.java:311)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.setup(ProcessLauncherDaemon.java:1177)
        at io.skymind.skil.daemon.ServiceRegister.checkDefaultModelHistoryServer(ServiceRegister.java:705)
        at io.skymind.skil.daemon.ServiceRegister.access$300(ServiceRegister.java:47)
        at io.skymind.skil.daemon.ServiceRegister$1.takeLeadership(ServiceRegister.java:145)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$WrappedListener.takeLeadership(LeaderSelector.java:536)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.doWork(LeaderSelector.java:399)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.doWorkLoop(LeaderSelector.java:443)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.access$100(LeaderSelector.java:64)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$2.call(LeaderSelector.java:245)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$2.call(LeaderSelector.java:239)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Cannot initialize Hdfs2ModelStorage: HDFS path (path to HDFS cluster namenode) not setCannot initialize Hdfs2ModelStorage: HDFS path (path to HDFS cluster namenode) not set

**Comment by wangfeng-skymind** [comment]
@wmeddie what cause is this?

**Comment by wmeddie** [comment]
Is this the RPM or docker?

There's also two skil-env.sh.  Did you add it to /etc/skil/skil-env.sh?

**Comment by wangfeng-skymind** [comment]
@wmeddie 
i had set these arguments in the /etc/skil/skil-env.sh,not /etc/profile.d/skil-env.sh

**Comment by wangfeng-skymind** [comment]
using  yum install skil-server skil-server-miniconda skil-server-spark

**Comment by wmeddie** [comment]
Can you try removing the skil.* options and instead put them together in JAVA_TOOL_OPTIONS like this:

```
    JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000-D
skil.hdfs2modelstorage.basedir=/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"
```

**Comment by wangfeng-skymind** [comment]
**i put these argument**  JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -D
skil.hdfs2modelstorage.basedir=/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root" **to /etc/skil/skil-env.sh,** 

**throw other exception:**
Picked up JAVA_TOOL_OPTIONS: MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage -Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -D skil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root
Unrecognized option: MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
@wmeddie

**Comment by wmeddie** [comment]
Looks like the JAVA_TOOL_OPTIONS is not formatted correctly.  Can you Gist the entire contents of /etc/skil/skil-env.sh?

**Comment by wangfeng-skymind** [comment]
SKIL_HOME=/opt/skil
SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/etc/skil/*
SKIL_LOG_DIR=/var/log/skil
SKIL_PID_FILE=/run/skil/skil.pid

SKIL_BACKEND=cpu
SKIL_LICENSE_PATH=/etc/skil/license.txt
SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt

JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
JDK_HOME=$JAVA_HOME

SKIL_SERVER_PROD_MODE=true

ZOOKEEPER_HOST=skil1:2181,skil2:2181
ZOOKEEPER_EMBEDDED=false


SKIL_USE_EMBEDDED_DB=false
SKIL_DB_DRIVER=com.mysql.jdbc.Driver
SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil
SKIL_DB_USER=skil
SKIL_DB_PASSWORD=skil


JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -D skil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"

**Comment by wmeddie** [comment]
What happened to MODEL_STORAGE_CLASS?

**Comment by wangfeng-skymind** [comment]
Picked up JAVA_TOOL_OPTIONS: -Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -D skil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hd
fs2modelstorage.hadoopuser=root
Unrecognized option: skil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.

**Comment by wmeddie** [comment]
JAVA_TOOL_OPTIONS should all be a single line:

```
JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"
```

**Comment by wangfeng-skymind** [comment]
previous exception :it's JAVA_TOOL_OPTIONS="MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage -Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -D skil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"
~

**Comment by wangfeng-skymind** [comment]
let me try again,thanks

**Comment by wmeddie** [comment]
It should look like:

```
SKIL_HOME=/opt/skil
SKIL_CLASS_PATH=/opt/skil/lib/:/opt/skil/native/:/opt/skil/jackson-2.5.1/:/etc/skil/
SKIL_LOG_DIR=/var/log/skil
SKIL_PID_FILE=/run/skil/skil.pid

SKIL_BACKEND=cpu
SKIL_LICENSE_PATH=/etc/skil/license.txt
SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt

JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
JDK_HOME=$JAVA_HOME

SKIL_SERVER_PROD_MODE=true

ZOOKEEPER_HOST=skil1:2181,skil2:2181
ZOOKEEPER_EMBEDDED=false

SKIL_USE_EMBEDDED_DB=false
SKIL_DB_DRIVER=com.mysql.jdbc.Driver
SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil
SKIL_DB_USER=skil
SKIL_DB_PASSWORD=skil

MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage

JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"
```

**Comment by wangfeng-skymind** [comment]
**Seem to be new exception:
/etc/skil/skil-env.sh 's content**
SKIL_HOME=/opt/skil
SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/etc/skil/*
SKIL_LOG_DIR=/var/log/skil
SKIL_PID_FILE=/run/skil/skil.pid

SKIL_BACKEND=cpu
SKIL_LICENSE_PATH=/etc/skil/license.txt
SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt

JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
JDK_HOME=$JAVA_HOME

SKIL_SERVER_PROD_MODE=true

ZOOKEEPER_HOST=skil1:2181,skil2:2181
ZOOKEEPER_EMBEDDED=false


SKIL_USE_EMBEDDED_DB=false
SKIL_DB_DRIVER=com.mysql.jdbc.Driver
SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil
SKIL_DB_USER=skil
SKIL_DB_PASSWORD=skil


MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage
JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://192.168.122.138:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://192.168.122.138:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"

**throw exception info:**
IL_HOME=/opt/skil, SKIL_DB_USER=skil, SHLVL=2, HOME=/opt/skil}
09:49:16.989 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Initializing Hdfs2ModelStorage with values:
09:49:16.989 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hdfsuri = hdfs://192.168.122.138:9000
09:49:16.989 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.basedir = hdfs://192.168.122.138:9000/user/hadoop/modelstorage
09:49:16.989 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hadoopuser = root
09:49:16.992 [main] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage)
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.runMain(ProcessLauncherDaemon.java:522)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.main(ProcessLauncherDaemon.java:1394)
Caused by: java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
        at io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage.<init>(Hdfs2ModelStorage.java:84)
        ... 6 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 7 common frames omitted
09:49:17.013 [Curator-LeaderSelector-0] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error starting process
java.lang.InterruptedException: null
        at java.lang.Object.wait(Native Method)
        at java.lang.Object.wait(Object.java:502)
        at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1342)
        at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:781)
        at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:691)
        at org.apache.curator.framework.imps.CreateBuilderImpl$11.call(CreateBuilderImpl.java:675)
        at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:107)
        at org.apache.curator.framework.imps.CreateBuilderImpl.pathInForeground(CreateBuilderImpl.java:672)
        at org.apache.curator.framework.imps.CreateBuilderImpl.protectedPathInForeground(CreateBuilderImpl.java:453)
        at org.apache.curator.framework.imps.CreateBuilderImpl.forPath(CreateBuilderImpl.java:443)
        at org.apache.curator.framework.imps.CreateBuilderImpl$3.forPath(CreateBuilderImpl.java:251)
        at org.apache.curator.framework.imps.CreateBuilderImpl$3.forPath(CreateBuilderImpl.java:205)
        at io.skymind.skil.daemon.ServiceRegister.register(ServiceRegister.java:436)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.registerService(ProcessLauncherDaemon.java:311)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.setup(ProcessLauncherDaemon.java:1177)
        at io.skymind.skil.daemon.ServiceRegister.checkDefaultModelHistoryServer(ServiceRegister.java:705)
        at io.skymind.skil.daemon.ServiceRegister.access$300(ServiceRegister.java:47)
        at io.skymind.skil.daemon.ServiceRegister$1.takeLeadership(ServiceRegister.java:145)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$WrappedListener.takeLeadership(LeaderSelector.java:536)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.doWork(LeaderSelector.java:399)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.doWorkLoop(LeaderSelector.java:443)
        at org.apache.curator.framework.recipes.leader.LeaderSelector.access$100(LeaderSelector.java:64)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$2.call(LeaderSelector.java:245)
        at org.apache.curator.framework.recipes.leader.LeaderSelector$2.call(LeaderSelector.java:239)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

**Comment by wmeddie** [comment]
Ah, yes, and change the SKIL_CLASS_PATH to:

```
SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/opt/skil/hadoop-common/*:/etc/skil/*
```

**Comment by wangfeng-skymind** [comment]
hardware structure: in total 4 devices, their domain  is skil1,skil2, hdfs1,hdfs2, the hdfs1 and hdfs2 make up distribution file cluster(HDFS),hdfs1 is master node
the domain skil1 and skil2 install skil cluster and zookeeper cluster, but mysql only was installed to skil1;

**below the /etc/skil/skil-evn.sh content:**
SKIL_HOME=/opt/skil
SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/etc/skil/*
SKIL_LOG_DIR=/var/log/skil
SKIL_PID_FILE=/run/skil/skil.pid
SKIL_BACKEND=cpu
SKIL_LICENSE_PATH=/etc/skil/license.txt
SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt
JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
JDK_HOME=$JAVA_HOME
SKIL_SERVER_PROD_MODE=true
ZOOKEEPER_HOST=skil1:2181,skil2:2181
ZOOKEEPER_EMBEDDED=false
SKIL_USE_EMBEDDED_DB=false
SKIL_DB_DRIVER=com.mysql.jdbc.Driver
SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil
SKIL_DB_USER=skil
SKIL_DB_PASSWORD=skil
MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage
JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://hdfs1:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://hdfs1:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"

throw exception info:
17:33:19.278 [main] INFO  org.eclipse.jetty.server.Server - Started @3048ms
17:33:19.299 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Initializing Hdfs2ModelStorage with values:
17:33:19.299 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hdfsuri = hdfs://hdfs1:9000
17:33:19.299 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.basedir = hdfs://hdfs1:9000/user/hadoop/modelstorage
17:33:19.299 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hadoopuser = root
17:33:19.302 [main] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage)
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.runMain(ProcessLauncherDaemon.java:522)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.main(ProcessLauncherDaemon.java:1394)
Caused by: java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
        at io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage.<init>(Hdfs2ModelStorage.java:84)
        ... 6 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 7 common frames omitted

**Comment by wmeddie** [comment]
Did you add the `/opt/skil/hadoop-common/*` to the SKIL_CLASS_PATH?

**Comment by wangfeng-skymind** [comment]
yeah, currently had add /opt/skil/hadoop-common/* to /etc/skil/skil-env.sh, however throw same exception, and then i find the directory which is /opt/skil/hadoop-common don‘t exist

[root@skil1 skil]# ll
total 56
-rw-r--r--  1 skil skil    36 Jul 28 15:06 agentid
drwxr-xr-x  2 skil skil   175 Jul 24 12:59 cuda
drwxr-xr-x  2 skil skil    50 Jul 24 12:59 docker
drwxr-xr-x  2 skil skil   230 Jul 24 12:59 jackson-2.4.4
drwxr-xr-x  2 skil skil   230 Jul 24 12:59 jackson-2.5.1
drwxr-xr-x  2 skil skil    33 Jul 24 12:59 jersey-1.9
drwxr-xr-x  2 skil skil    69 Jul 24 12:59 jersey-2.0.1
drwxr-xr-x  3 skil skil 32768 Jul 26 22:00 lib
drwxr-xr-x 21 root root   308 Jul 24 13:03 miniconda
drwxr-xr-x  2 skil skil   125 Jul 24 12:59 native
drwxr-xr-x  4 skil skil    31 Jul 28 15:06 plugins
drwxr-xr-x  5 skil skil   275 Jul 24 12:59 sbin
drwxr-xr-x  2 skil skil  4096 Jul 24 13:00 spark
[root@skil1 skil]#

**Comment by wangfeng-skymind** [comment]
i copy hadoop-common from other repository to /opt/skil/, but still same exception,

**Comment by wangfeng-skymind** [comment]
09:39:54.344 [main] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage)
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.runMain(ProcessLauncherDaemon.java:522)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.main(ProcessLauncherDaemon.java:1394)
Caused by: java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
        at io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage.<init>(Hdfs2ModelStorage.java:84)
        ... 6 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 7 common frames omitted

@wmeddie how to solve?

**Comment by wmeddie** [comment]
Can you try adding `/opt/skil/spark/*`to SKIL_CLASS_PATH instead?

While that should fix the problem, it might introduce others.  If so let me
know and I'll see exactly what jars we should add to SKIL_CLASS_PATH.

**Comment by wangfeng-skymind** [comment]
SKIL_HOME=/opt/skil
SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/opt/skil/hadoop-common/*:/opt/skil/spark/*:/etc/skil/*
SKIL_LOG_DIR=/var/log/skil
SKIL_PID_FILE=/run/skil/skil.pid
SKIL_BACKEND=cpu
SKIL_LICENSE_PATH=/etc/skil/license.txt
SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt
JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
JDK_HOME=$JAVA_HOME
SKIL_SERVER_PROD_MODE=true
ZOOKEEPER_HOST=skil1:2181,skil2:2181
ZOOKEEPER_EMBEDDED=false
SKIL_USE_EMBEDDED_DB=false
SKIL_DB_DRIVER=com.mysql.jdbc.Driver
SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil
SKIL_DB_USER=skil
SKIL_DB_PASSWORD=skil
MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage
JAVA_TOOL_OPTIONS="-Dskil.hdfs2modelstorage.hdfsuri=hdfs://hdfs1:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://hdfs1:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root"



1:38:36.554 [main] INFO  i.s.s.daemon.ProcessLauncherDaemon - Setting up plugin server /opt/skil/plugins with port 9508
21:38:36.594 [main] INFO  org.eclipse.jetty.util.log - Logging initialized @2756ms
21:38:36.655 [main] INFO  org.eclipse.jetty.server.Server - jetty-9.2.19.v20160908
21:38:36.677 [main] INFO  o.e.jetty.server.ServerConnector - Started ServerConnector@1f44ddab{HTTP/1.1}{0.0.0.0:9508}
21:38:36.678 [main] INFO  org.eclipse.jetty.server.Server - Started @2840ms
21:38:36.696 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Initializing Hdfs2ModelStorage with values:
21:38:36.696 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hdfsuri = hdfs://hdfs1:9000
21:38:36.696 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.basedir = hdfs://hdfs1:9000/user/hadoop/modelstorage
21:38:36.696 [main] INFO  i.s.m.hdfs2.Hdfs2ModelStorage - Hdfs2ModelStorage config: skil.hdfs2modelstorage.hadoopuser = root
21:38:36.699 [main] ERROR i.s.s.daemon.ProcessLauncherDaemon - Error initializing model storage: SKIL will now exit (model storage class: io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage)
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.runMain(ProcessLauncherDaemon.java:522)
        at io.skymind.skil.daemon.ProcessLauncherDaemon.main(ProcessLauncherDaemon.java:1394)
Caused by: java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
        at io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage.<init>(Hdfs2ModelStorage.java:84)
        ... 6 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 7 common frames omitted
21:38:36.705 [Curator-LeaderSelector-0] INFO  i.s.s.daemon.ProcessLauncherDaemon - Set jvm args to be -Xmx1g
21:38:36.714 [Curator-LeaderSelector-0] INFO  i.s.s.daemon.ProcessLauncherDaemon - Received args ModelHistoryArgs(super=BaseToArgs(id=7dd95426-045d-4ce1-ac9c-1615c6d97650, jvmArgs=-Xmx1g, agentId=null, name=Default Model History Server, metaData=null), dbPath=file:/var/skil/mh/default, dbName=xdb, userName=SA, password=, modelHistoryPort=9100, dbPort=9305, dbUrl=jdbc:hsqldb:file:/var/skil/mh/default) with environment to be {PATH=/opt/skil/miniconda/envs/skil-python-environment/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin, SKIL_CLASS_PATH=/opt/skil/lib/*:/opt/skil/native/*:/opt/skil/jackson-2.5.1/*:/opt/skil/hadoop-common/*:/opt/skil/spark/*:/etc/skil/*, CONDA_DEFAULT_ENV=skil-python-environment, JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk, LANG=en_US.UTF-8, SKIL_USE_EMBEDDED_DB=false, CONDA_PS1_BACKUP=, SKIL_PID_FILE=/run/skil/skil.pid, CONDA_PREFIX=/opt/skil/miniconda/envs/skil-python-environment, SKIL_LICENSE_PATH=/etc/skil/license.txt, ZOOKEEPER_EMBEDDED=false, SPARK_HOME=/opt/spark, JAVA_TOOL_OPTIONS=-Dskil.hdfs2modelstorage.hdfsuri=hdfs://hdfs1:9000 -Dskil.hdfs2modelstorage.basedir=hdfs://hdfs1:9000/user/hadoop/modelstorage -Dskil.hdfs2modelstorage.hadoopuser=root, SKIL_PUBLIC_KEY_PATH=/etc/skil/publickey.txt, LOGNAME=skil, JDK_HOME=$JAVA_HOME, SKIL_SERVER_PROD_MODE=true, PWD=/, SKIL_DB_URL=jdbc:mysql://192.168.122.128:3306/skil, SKIL_DB_PASSWORD=skil, _=/usr/bin/java, SHELL=/sbin/nologin, SKIL_LOG_DIR=/var/log/skil, CONDA_PATH_BACKUP=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin, MODEL_STORAGE_CLASS=io.skymind.modelstorage.hdfs2.Hdfs2ModelStorage, USER=skil, SKIL_DB_DRIVER=com.mysql.jdbc.Driver, SKIL_BACKEND=cpu, ZOOKEEPER_HOST=skil1:2181,skil2:2181, SKIL_HOME=/opt/skil, SKIL_DB_USER=skil, SHLVL=2, HOME=/opt/skil}

**Comment by wangfeng-skymind** [comment]
@wmeddie Still seem to exist this problem: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration

**Comment by wmeddie** [comment]
Ok. This looks like a bug in 1.1. We’ll try to fix it in the next bug fix release.

**Comment by ShamsUlAzeem** [comment]
@wmeddie was this issue resolved?
